### ISSUE TITLE
GE Scraper numbers are off, PDF more up to date better

### ISSUE THREAD

**Issue by zdavatz** [openZH/covid_19#327]
The number for the GE Scraper is off. We need to parse the PDF for Geneva to get the correct numbers from the scraper.

This is the correct URL for Geneva: https://www.ge.ch/document/covid-19-situation-epidemiologique-geneve/telecharger

cc @baryluk   

related. #323

**Comment by calmyournerves** [comment]
It looks like the table on top in the PDF is an embedded image. This makes it very hard to parse.

**Comment by metaodi** [comment]
Note that we already parse a PDF for GE, and we use pdftotext to extract the content.

**Comment by calmyournerves** [comment]
@metaodi Yeah, saw that. But `pdftotext` does not work for images. Here's the `pdftotext`-Output of the above URL:

```
REPUBLIQUE ET CANTON DE GENEVE
Département de la sécurité, de l'emploi et de la santé

Direction générale de la santé
Service du médecin cantonal

COVID-19 - Situation épidémiologique, canton de Genève
Situation au 29.03.2020 – 12h00, selon les données disponibles

Service du médecin cantonal • Rue Adrien-Lachenal 8 • 1207 Genève
Tél. +41 (0) 22 546 50 00 • Fax +41 (0) 22 546 97 51 • www.ge.ch
Lignes TPG 1, 12, 25, 61 - arrêt Terrassière


Page : 2/2
```

**Comment by andreasamsler** [comment]
Note that "we need to ..." is not what we need to do. We are grateful for the issues, everyone is reporting (except issues which are about bullying other contributors or actors in the current situation.

What we need to do, is to analyse each of the issues, you are reporting. And come up with solutions. We do not need to be fast primarily, but responsible with our actions. Thanks!

**Comment by zdavatz** [comment]
I agree, the numbers have to be correct. Maybe @baryluk has a good solution for scraping, using OCR ;).

**Comment by baryluk** [comment]
Yeah. We are using https://www.ge.ch/document/point-coronavirus-maladie-covid-19/telecharger at the moment, but last time it was updated on 2020-03-27.

https://www.ge.ch/document/covid-19-situation-epidemiologique-geneve/telecharger  has more information, but it has data as images, not something easy to extract.

**Comment by zdavatz** [comment]
I wanna try this tomorrow. https://medium.com/@MicroPyramid/extract-text-with-ocr-for-all-image-types-in-python-using-pytesseract-ec3c53e5fc3a

**Comment by baryluk** [comment]
@zdavatz Yeah. tesseract should be able to handle it.

I think tho, contacting people responsible for this data in Geneva to make it easier for us and others, is a better option :)

**Comment by zdavatz** [comment]
True. Will try that as well.

**Comment by zdavatz** [comment]
Actually [pytesseract](https://pypi.org/project/pytesseract/) works quite well, I am getting these numbers:

```
29.03 | 2349 92 | 27 59 | 55 6 279 365 44 193 37 5
30.03 | 2375 provisoire 26 provisoire 24 57 | 54 2 301 382 40 216 42 5
```

@baryluk I think you should be able to grab those, once you have the image from the PDF file.

**Comment by baryluk** [comment]
Looks very promising.

Images should be easy to extract using `pdfimages` from `popller-utils` (same package that provides `pdftotext`).

Sad news is that images are broken up into sub-parts partially:

1)

![-000](https://user-images.githubusercontent.com/110322/78067151-c3244480-7385-11ea-8b69-e061de36c06d.png)

2)

![-001](https://user-images.githubusercontent.com/110322/78067154-c4557180-7385-11ea-86a2-2046dc8ea59c.png)

Same with graphs.


Still salvageable.

**Comment by baryluk** [comment]
I tried tesseract on the second image, and got this:

```
$ tesseract -l fra --psm 11 ./-001.png stdout
PORC RIENIONCT

AVR

PILCNIDNES RUE

soins intensifs

URRE VIE JUNE

LU VILTLZ

Ur HIVSPVILONISOLIUEE

hospitalisés

30.03

250 À 26 À 2 À 57 À Sa 2 À 301 | 382 | 40 À 216 | 48 | 5

31.03

2451 | 1 À 20 À 64 À à À À 821 À 40 | 28 | 2412 | 49 | 6


```

That is not good enough. 250 is wrong, should be 2450.

26 is wrong, it should be 165.

2412 is wrong, should 241.

821 is wrong, it should be 321.

@zdavatz Could you provide more details how you run tesseract, with options, and upload the exact input that you gave it?

I am using tesseract 4.1.1, with fra, eng, and osd files at version 1:4.00~git30-7274cfa-1.

**Comment by zdavatz** [comment]
@baryluk did you install `pytesseract`with `pip install --user pytesseract` (on Linux). Once that is installed you can just do: `pytesseract image.png` and grab the data from there. See: https://pypi.org/project/pytesseract/

I think this task is made for you ;).

**Comment by baryluk** [comment]
@zdavatz I need to know more. Like exact version of libraries, tesseract and language files.

**Comment by baryluk** [comment]
@zdavatz 

I installed pytesseract and it doesn't work:

```
$ ./.local/bin/pytesseract 78067154-c4557180-7385-11ea-86a2-2046dc8ea59c.png
ERROR: Could not open file "78067154-c4557180-7385-11ea-86a2-2046dc8ea59c.png"
user@debian:~$ ls -l 78067154-c4557180-7385-11ea-86a2-2046dc8ea59c.png
-rw-r--r-- 1 user user 269736 Mar 31 19:28 78067154-c4557180-7385-11ea-86a2-2046dc8ea59c.png
$ file 78067154-c4557180-7385-11ea-86a2-2046dc8ea59c.png
78067154-c4557180-7385-11ea-86a2-2046dc8ea59c.png: PNG image data, 4334 x 196, 8-bit/color RGB, non-interlaced
$
```

**Comment by zdavatz** [comment]
All I really done is `pip install --user pytesseract` nothing else.  I also tested on my Mac with Python 2.7.17. 

I just now I tested with this image: 
<img width="764" alt="Bildschirmfoto 2020-03-31 um 21 44 04" src="https://user-images.githubusercontent.com/4953/78068706-266ba200-7399-11ea-911f-9bec3a83f0bc.png">

doing 

<img width="1096" alt="Bildschirmfoto 2020-03-31 um 21 47 38" src="https://user-images.githubusercontent.com/4953/78068746-3edbbc80-7399-11ea-8501-fcf3689194fe.png">

**Comment by zdavatz** [comment]
> @zdavatz
> 
> I installed pytesseract (0.3.3) and it doesn't work:
> 
> ```
> $ ./.local/bin/pytesseract 78067154-c4557180-7385-11ea-86a2-2046dc8ea59c.png
> ERROR: Could not open file "78067154-c4557180-7385-11ea-86a2-2046dc8ea59c.png"
> user@debian:~$ ls -l 78067154-c4557180-7385-11ea-86a2-2046dc8ea59c.png
> -rw-r--r-- 1 user user 269736 Mar 31 19:28 78067154-c4557180-7385-11ea-86a2-2046dc8ea59c.png
> $ file 78067154-c4557180-7385-11ea-86a2-2046dc8ea59c.png
> 78067154-c4557180-7385-11ea-86a2-2046dc8ea59c.png: PNG image data, 4334 x 196, 8-bit/color RGB, non-interlaced
> $
> ```

How did you take the Screenshot? I done it with `Command+Shift+4` on my Mac.

**Comment by baryluk** [comment]
@zdavatz How did you get an input image? Just screenshot it? `pdfimages`? `pdftoppm -png`?

**Comment by baryluk** [comment]
@zdavatz I used `pdfimages` to exact raw embedded images directly from the PDF without any conversion or resizing.

**Comment by zdavatz** [comment]
See above, just posted. On my Mac I done `Command+Shift+4" and fed that to pytesseract. What version of Python are you using?

**Comment by zdavatz** [comment]
> @zdavatz I used `pdfimages` to exact raw embedded images directly from the PDF without any conversion or resizing.

Can you try grabing the first 30% of the first page using imagemagick?

**Comment by zdavatz** [comment]
The image you posted above does not give me good results with `pytesseract`. Must be something with the way you take the image.

**Comment by baryluk** [comment]
I tried `pytesseract` again, with some fixes, and it gives me exactly same results as using `tesseract-ocr` from command line. pytesseract is just a helper / library that utilized `tesseract-ocr`, so there will be no quality changes using one or another.

**Comment by baryluk** [comment]
@zdavatz Well, I will check.

I tried this: 
![telecharger](https://user-images.githubusercontent.com/110322/78070084-b8b87980-738a-11ea-91ef-990c5747a5b5.png)

And it makes a big mess:

```
$ tesseract -l fra ./telecharger.png stdout
Warning: Invalid resolution 0 dpi. Using 70 instead.
Estimating resolution as 242
REPUBLIQUE ET CANTON DE GENEVE
Département de la sécurité, de l'emploi et de la santé

Direction générale de la santé
Service du médecin cantonal

   

ART TER EANAS LU

COVID-19 - Situation épidémiologique, canton de Genève
Situation au 31.03.2020 — 12h00, selon les données disponibles

 

2451 provioïe 1 proviore 20 [688 1 522 7 20 7 28 7 22 7 40 1 6

  

 

 

       

 

 

 

Nombre de tests et de cas COVID-19, canton de Genève Nombre de cas COVID-19 hospitalisés, canton de Genève
450
800 70
400
700 ma Nombre cas COVID-19 60 m Patients COVID-19 hospitalisés en
= 350 ï ï
2 1 Nombre tests négatifs & unité de soins
600 mn :
ê —% tests positifs 50 £ 300 B Patients COVID-19
S% 500 ‘5 8 aux soins intermédiaires
&
£ 40 8 3 250 m Patients COVID-19
5 400 £ £ 200 aux soins intensifs total
Z &
30 à Ë
300 È z 150
20
200 5 100
à
100 10 50
2
0 le 0 0
838282832382 8282 8 2 8 8 N À ON 9 m mm nm m m m mm m m m m mm m m
AR A S 8 8 S S S SE 8 N * 8 D RD RSS S SN Sd 8 9 À à 8 à ©
ë D LD] D] e o eo + Le Le Le LS LI Li LS LU a m
date ë date
x Cas COVID-19 hospitalisés aux soins intensifs, canton de Genève
3000 Cumul des cas COVID-19, canton de Genève 70 P !
60 m Patients COVID-19
2500 5 4 315
aux soins intensifs intubés
50 .
8 2000 u Patients COVID-19 .
2 8 aux soins intensifs non intubés
s & 40
® TT
Ê 1500 o
2
L Ë 30
=
1000
20
500 10
0 o
14587898 8 4 8 8 8 8 8 8 8
4 $ à S à à ë R #8 À 8 8 8 8
:
Date ê Date
Nombre de nouveaux patients COVID-19 hospitalisés, admis aux soins
60 Cumul des décès COVID-19, canton de Genève intensifs et décédés, canton de Genève
60
50 Em Nb nouveaux patients COVID-
50 19 hospitalisés 50 e
8 Em Nb nouveaux décès COVID-19 4
40 x 20
8 3 40 j #
So ÿ —-Nb nouveaux patients COVID- 36
5 3 19 admis aux soins intensifs
o 30 =
£ 30 28
5 24
20
20
10
10
0
NS ON na m m æm mi 29 9 9m m nm m m a 0
S S $$$ 22 2 2 2 2 2 2 2 S L 2 © 7 7 r 7 r 7 '
À $ À 8 8 8 8 3 4 4 2 # 8 N à À À & 8 8 8 8 8 8 8 8 8 8 8 8
Date sS < LI mn s ñ os = a a sS s
& à à à à à k à % ñ 8 ñ
date

 

 

 

Service du médecin cantonal - Rue Adrien-Lachenal 8 + 1207 Genève
Tél. +41 (0) 22 546 50 00 + Fax +41 (0) 22 546 97 51 + www.ge.ch
Lignes TPG 1, 12, 25, 61 - arrêt Terrassière
$
```

with `--psm 11`:

```
$ tesseract -l fra --psm 11 ./telecharger.png stdout
Warning: Invalid resolution 0 dpi. Using 70 instead.
Estimating resolution as 242
EURE

REPUBLIQUE ET CANTON DE GENEVE

FE

Département de la sécurité, de l'emploi et de la santé

LS

sara

2

Direction générale de la santé

Service du médecin cantonal

COVID-

9 - Situation épidémiologique, canton de Genève

Situation au 31.03.2020 — 12h00, selon les données disponibles

Cumul COVID-19

uv

sorties

©

d'hospitalisation

1

30.03

AE PRE PE RE PE PE EE EE RE OR RE ES ES A ES

31.03

[2451 provisoire 1° provisoire 20] 641 80 1 522] 208 7 28 7 21 7 40 1 6
...

```

Again mess and mistakes.

It looks like very sensitive to other stuff on page.

In the mean time I filled this: https://github.com/madmaze/pytesseract/issues/264 to tesseract team.

**Comment by zdavatz** [comment]
@baryluk The nice thing about pytesseract is, that it uses the Google OCR API. I think key is that you first take a picture of the first 30% of the image using imagemagick and then you feed that to pytesseract.

Tesseract in itself is a beast. We stopped using it once Google/Apple came out with their Vision APIs.

**Comment by zdavatz** [comment]
I wann try this tomorrow with [imagemagick](http://www.imagemagick.org/discourse-server/viewtopic.php?t=35073): `convert 'myfile.pdf[0]' outfile.jpg` then feed that to `pytesseract`.

**Comment by baryluk** [comment]
@zdavatz I tried with convert already. It results in same issues, or is even worse.

I used `convert -density 300 input.pdf output.png`

**Comment by baryluk** [comment]
I am testing https://cloud.google.com/vision/docs/ocr and  https://cloud.google.com/vision/docs/pdf now.

**Comment by baryluk** [comment]
@zdavatz pytesseract doesn't use Google OCR APIs. If I don't have tesseract installed locally it fails to even start.

**Comment by baryluk** [comment]
This is result of feeding just the first table to Google Vision annotate API:

[table1.json.txt](https://github.com/openZH/covid_19/files/4411487/table1.json.txt)

It appears somehow usable, but still requires a lot of postprocessing, sorting to do a layout, etc.

**Comment by baryluk** [comment]
FYI. https://www.ge.ch/document/point-coronavirus-maladie-covid-19/telecharger was just updated, and numbers looks good:

```
$ ./scrape_ge.sh 
GE
Downloading: https://www.ge.ch/document/point-coronavirus-maladie-covid-19/telecharger
Scraped at: 2020-04-01T04:37:45+02:00
Date and time: 31.03 à 12h
Confirmed cases: 2451
Hospitalized: 405
ICU: 64
Deaths: 49
$
```

**Comment by zdavatz** [comment]
great!

**Comment by zdavatz** [comment]
cc @baryluk this works for me:

1. `gs -o cropped.pdf -sDEVICE=pdfwrite -c "[/CropBox [24 600 800 650]" -c "/PAGES pdfmark" -f covid-19_ge_situation_01_04_20-s.pdf`
2. `convert -density 500 cropped.pdf baryluk.png`
3. 
```/tmp> pytesseract baryluk.png 
2702 provisoire 45 provisoire
```
